junit5/src/main/java/org/jboss/weld/junit5/WeldJunit5Extension.java

```diff
@@ -196,85 +198,106 @@ private boolean methodRequiresExplicitParamInjection(ParameterContext pc) {
-                                    + "in class " + field.getDeclaringClass());
-                        }
-                    }
-                }
-            }
-            // Multiple occurrences of @WeldSetup in the hierarchy will lead to an exception
-            if (foundInitiatorFields.size() > 1) {
-                throw new IllegalStateException(foundInitiatorFields.stream().map(f -> "Field type - " + f.getType() + " which is "
-                        + "in " + f.getDeclaringClass()).collect(Collectors.joining("\n", "Multiple @WeldSetup annotated fields found, "
-                        + "only one is allowed! Fields found:\n", "")));
-            }
+            // iterate through the testInstance, the enclosing instance (in case of nested tests),
+            // the enclosing instance of the enclosing instance (in cases of twice nested tests) and so on
+            // until we find a WeldInitiator
+            final List<Object> allTestInstances = new ArrayList<>(context.getRequiredTestInstances().getAllInstances());
+            Collections.reverse(allTestInstances); // so we can iterate from inner-most to outer-most
+            WeldInitiator initiator = allTestInstances.stream()
+                    .map(this::findInitiatorInInstance)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElseGet(() -> getDefaultInitiator(context, testInstance));
+            setInitiatorToStore(context, initiator);
 
-            // at this point we can be sure that either no or exactly one WeldInitiator was found
-            if (initiator == null) {
-                Weld weld = WeldInitiator.createWeld();
-                WeldInitiator.Builder builder = WeldInitiator.from(weld);
+            // this ensures the test class is injected into
+            // in case of nested tests, this also injects into any outer classes
+            initiator.addObjectsToInjectInto(new HashSet<>(allTestInstances));
 
-                weldInit(context, weld, builder);
+            // and finally, init Weld
+            setContainerToStore(context, initiator.initWeld(testInstance));
+        }
+    }
 
-                // Apply discovered enrichers
-                for (WeldJunitEnricher enricher : getEnrichersFromStore(context)) {
-                    String property = System.getProperty(enricher.getClass().getName());
-                    if (property == null || Boolean.parseBoolean(property)) {
-                        enricher.enrich(testInstance, context, weld, builder);
+    private WeldInitiator findInitiatorInInstance(Object testInstance) {
+        // all found fields which are WeldInitiator and have @WeldSetup annotation
+        List<Field> foundInitiatorFields = new ArrayList<>();
+        // We will go through class hierarchy in search of @WeldSetup field (even private)
+        for (Class<?> clazz = testInstance.getClass(); clazz != null; clazz = clazz.getSuperclass()) {
```

Comment:
We ought to have a test that verifies that superclass configuration takes precedence over outer class config for a `@Nested` test that has no config of its own.